File: feldspar/MarketDataService/Providers/AlphaVantage.cs
namespace MarketDataService.Providers;

public class AlphaVantage
{
    private readonly Uri _baseUri;

    public AlphaVantage(string baseUri, string apiKey)
    {
        if (baseUri == null) throw new ArgumentNullException(nameof(baseUri));
        if (string.IsNullOrEmpty(apiKey))
        {
            throw new ArgumentException($"'{nameof(apiKey)}' cannot be null or empty.", nameof(apiKey));
        }

        _baseUri = new Uri($"{baseUri}/query?apikey={apiKey}");
    }

    private Uri CommodityQuery(string commodity, string interval)
    {
        var uriString = $"{_baseUri}&function={commodity}&interval={interval}";
        return new Uri(uriString);
    }

    public async Task<string> GetCommodityData(string commodity, string interval)
    {
        var queryUri = CommodityQuery(commodity, interval);
        using var client = new HttpClient();
        var response = await client.GetAsync(queryUri);
        response.EnsureSuccessStatusCode();
        var responseString = await response.Content.ReadAsStringAsync();
        return responseString;
    }
}

File: feldspar/jMarketDataService/src/main/java/com/feldspar/MarketDataService/AlphaVantage.java
package com.feldspar.MarketDataService;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

public class AlphaVantage {

    private final URI baseUri;

    public AlphaVantage(String baseUri, String apiKey) throws URISyntaxException {

        this.baseUri = new URI(String.format("%s/query?apikey=%s", baseUri, apiKey));
    }

    private URI CommodityQuery(String commodity, String interval) throws URISyntaxException
    {
        var uriString = String.format("%s&function=%s&interval=%s", baseUri, commodity, interval);
        return new URI(uriString);
    }

    public String GetCommodityData(String name, String interval) throws IOException, InterruptedException, URISyntaxException
    {
        var queryUri = CommodityQuery(name, interval);
        var client = HttpClient.newHttpClient();
        var request = HttpRequest.newBuilder(queryUri).build();
        var response = client.send(request, HttpResponse.BodyHandlers.ofString());
        return response.body();
    }
}
